karapace/protobuf/protobuf_to_dict.py

```diff
@@ -0,0 +1,338 @@
+                pb, value_field, type_callable_map, use_enum_labels, including_default_value_fields, lowercase_enum_lables
+            )
+            for k, v in value.items():
+                result_dict[field.name][k] = type_callable(v)
+            continue
+        type_callable = _get_field_value_adaptor(
+            pb, field, type_callable_map, use_enum_labels, including_default_value_fields, lowercase_enum_lables
+        )
+        if field.label == FieldDescriptor.LABEL_REPEATED:
+            type_callable = repeated(type_callable)
+
+        if field.is_extension:
+            extensions[str(field.number)] = type_callable(value)
+            continue
+
+        result_dict[field.name] = type_callable(value)
+
+    # Serialize default value if including_default_value_fields is True.
+    if including_default_value_fields:
+        for field in pb.DESCRIPTOR.fields:
+            # Singular message fields and oneof fields will not be affected.
+            if (field.label != FieldDescriptor.LABEL_REPEATED and field.cpp_type == FieldDescriptor.CPPTYPE_MESSAGE):
+                continue
+            if field.containing_oneof:
+                continue
+            if field.name in result_dict:
+                # Skip the field which has been serailized already.
+                continue
+            if _is_map_entry(field):
+                result_dict[field.name] = {}
+            elif field.label == FieldDescriptor.LABEL_REPEATED:
+                result_dict[field.name] = []
+            elif field.type == FieldDescriptor.TYPE_ENUM and use_enum_labels:
+                result_dict[field.name] = enum_label_name(field, field.default_value, lowercase_enum_lables)
+            else:
+                result_dict[field.name] = field.default_value
+
+    if extensions:
+        result_dict[EXTENSION_CONTAINER] = extensions
+    return result_dict
+
+
+def _get_field_value_adaptor(
+    pb,
+    field,
+    type_callable_map=TYPE_CALLABLE_MAP,
+    use_enum_labels=False,
+    including_default_value_fields=False,
+    lowercase_enum_lables=False
+):
+
+    if field.message_type and field.message_type.name == Timestamp_type_name:
+        return timestamp_to_datetime
+    if field.type == FieldDescriptor.TYPE_MESSAGE:
+        # recursively encode protobuf sub-message
+        return lambda pb: protobuf_to_dict(
+            pb,
+            use_enum_labels=use_enum_labels,
+            including_default_value_fields=including_default_value_fields,
+            lowercase_enum_lables=lowercase_enum_lables,
+        )
+
+    if use_enum_labels and field.type == FieldDescriptor.TYPE_ENUM:
+        return lambda value: enum_label_name(field, value, lowercase_enum_lables)
+
+    if field.type in type_callable_map:
+        return type_callable_map[field.type]
+
+    raise TypeError("Field %s.%s has unrecognised type id %d" % (pb.__class__.__name__, field.name, field.type))
+
+
+REVERSE_TYPE_CALLABLE_MAP = frozendict({})
```

Comment:
```suggestion
REVERSE_TYPE_CALLABLE_MAP = MappingProxyType({})
```

Is this really supposed to be empty?

Reply:
yes